be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp

```diff
@@ -0,0 +1,228 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/pipeline/hashjoin/spillable_hash_join_build_operator.h"
+
+#include <atomic>
+#include <memory>
+
+#include "column/column_helper.h"
+#include "column/vectorized_fwd.h"
+#include "common/statusor.h"
+#include "exec/hash_join_node.h"
+#include "exec/join_hash_map.h"
+#include "exec/pipeline/hashjoin/hash_join_build_operator.h"
+#include "exec/pipeline/query_context.h"
+#include "exec/spill/options.h"
+#include "exec/spill/spiller.h"
+#include "exec/spill/spiller.hpp"
+#include "gen_cpp/InternalService_types.h"
+#include "gen_cpp/PlanNodes_types.h"
+#include "runtime/runtime_state.h"
+#include "util/defer_op.h"
+
+namespace starrocks::pipeline {
+
+Status SpillableHashJoinBuildOperator::prepare(RuntimeState* state) {
+    HashJoinBuildOperator::prepare(state);
+    _join_builder->spiller()->set_metrics(spill::SpillProcessMetrics(_unique_metrics.get()));
+    RETURN_IF_ERROR(_join_builder->spiller()->prepare(state));
+    if (state->spill_mode() == TSpillMode::FORCE) {
+        set_spill_strategy(spill::SpillStrategy::SPILL_ALL);
+    }
+    return Status::OK();
+}
+
+void SpillableHashJoinBuildOperator::close(RuntimeState* state) {
+    HashJoinBuildOperator::close(state);
+}
+
+bool SpillableHashJoinBuildOperator::need_input() const {
+    return !is_finished() && !(_join_builder->spiller()->is_full() || _join_builder->spill_channel()->has_task());
+}
+
+Status SpillableHashJoinBuildOperator::set_finishing(RuntimeState* state) {
+    if (spill_strategy() == spill::SpillStrategy::NO_SPILL) {
+        return HashJoinBuildOperator::set_finishing(state);
+    }
+
+    RETURN_IF_ERROR(publish_runtime_filters(state));
+
+    auto io_executor = _join_builder->spill_channel()->io_executor();
+    auto set_call_back_function = [this](RuntimeState* state, auto io_executor) {
+        _join_builder->spill_channel()->set_finishing();
+        return _join_builder->spiller()->set_flush_all_call_back(
+                [this]() {
+                    _is_finished = true;
+                    _join_builder->enter_probe_phase();
+                    return Status::OK();
+                },
+                state, *io_executor, spill::MemTrackerGuard(tls_mem_tracker));
+    };
+    RETURN_IF_ERROR(set_call_back_function(state, io_executor));
+
+    return Status::OK();
+}
+
+Status SpillableHashJoinBuildOperator::publish_runtime_filters(RuntimeState* state) {
+    // publish empty runtime filters
```

Comment:
the runtime filters may be not right if some build operators spill but others not?

Reply:
when we merge runtime filters. it will generate empty runtime filter